lib/fizzy/parser.cpp

```diff
@@ -298,19 +298,47 @@ Module parse(bytes_view input)
         }
 
         if (it != expected_end_pos)
+        {
             throw parser_error{"incorrect section " + std::to_string(static_cast<int>(id)) +
                                " size, difference: " + std::to_string(it - expected_end_pos)};
+        }
+    }
+
+    // Validation checks
+    if (module.tablesec.size() > 1)
+        throw parser_error{"too many table sections (at most one is allowed)"};
+
+    if (module.memorysec.size() > 1)
+        throw parser_error{"too many memory sections (at most one is allowed)"};
+
+    const auto imported_mem_count = std::count_if(module.importsec.begin(), module.importsec.end(),
+        [](const auto& import) noexcept { return import.kind == ExternalKind::Memory; });
 
-        if (module.tablesec.size() > 1)
-            throw parser_error{"too many table sections (at most one is allowed)"};
+    if (imported_mem_count > 1)
+        throw parser_error{"too many imported memories (at most one is allowed)"};
 
-        if (module.memorysec.size() > 1)
-            throw parser_error{"too many memory sections (at most one is allowed)"};
+    if (!module.memorysec.empty() && imported_mem_count > 0)
+    {
+        throw parser_error{
+            "both module memory and imported memory are defined (at most one of them is "
+            "allowed)"};
```

Comment:
Is this split due to clang-format?

Reply:
Actually looks like can fit into one line, will fix.